### PR TITLE
Added bot states and battery, refactored update methods.

### PR DESCRIPTION
**Pull request by RenierM26**
Hi @Danielhiversen, 

-Added switchbot bot state checking for on/off, battery and mode.
-Refactored the update methods. (Should allow more accurate updates in hassio.)
-Fixed Curtain battery check. 

The binary AND operators I found in the Official switchbot Node JS libraries. Tested them locally.

### CODE REVIEW

File: switchbot/__init__.py [hunk 1]
```diff
@@ -111,12 +115,41 @@ def _sendcommand(self, key, retry) -> bool:
...
+            time.sleep(waiting_time)
+        try:
+            devices = bluepy.btle.Scanner().scan(scan_timeout)
+
+        except bluepy.btle.BTLEManagementError:
+            _LOGGER.warning("Error updateing Switchbot.", exc_info=True)
+
+        if not devices:
+            time.sleep(DEFAULT_RETRY_TIMEOUT)
+            return self.get_servicedata(retry - 1, scan_timeout)
```
Comment (Danielhiversen) [inline switchbot/__init__.py:142]: This will retry forever?
Check if retry > 0

File: switchbot/__init__.py [hunk 2]
```diff
@@ -111,12 +115,41 @@ def _sendcommand(self, key, retry) -> bool:
...
+            )
             return False
-        _LOGGER.warning("Cannot connect to Switchbot. Retrying (remaining: %d)...", retry)
+        _LOGGER.warning(
+            "Cannot connect to Switchbot. Retrying (remaining: %d)...", retry
+        )
         time.sleep(DEFAULT_RETRY_TIMEOUT)
         return self._sendcommand(key, retry - 1)
 
+    def get_servicedata(self, retry=DEFAULT_RETRY_COUNT, scan_timeout=5) -> None:
```
Comment (Danielhiversen) [inline switchbot/__init__.py:128]: It does not return None

File: switchbot/__init__.py [hunk 3]
```diff
@@ -129,6 +162,20 @@ def get_min_time_update(self):
 class Switchbot(SwitchbotDevice):
     """Representation of a Switchbot."""
 
+    def __init__(self, *args, **kwargs) -> None:
+        self._battery_percent = 0
+        self._is_on = 0
+        self._mode = 0
```
Comment (Danielhiversen) [inline switchbot/__init__.py:168]: Maybe better to initialize as None?

File: switchbot/__init__.py [hunk 4]
```diff
@@ -141,6 +188,25 @@ def press(self) -> bool:
         """Press command to device."""
         return self._sendcommand(PRESS_KEY, self._retry_count)
 
+    def switch_mode(self) -> str:
+        """Return Toggle or Switch from cache.
+        Run update first."""
+        if self._mode != 0:
+            return "switch"
```
Comment (Danielhiversen) [inline switchbot/__init__.py:195]: I think it is better to move this logic to the update function

File: switchbot/__init__.py [hunk 5]
```diff
@@ -141,6 +188,25 @@ def press(self) -> bool:
...
+        return "toggle"
+
+    def is_on(self) -> bool:
+        """Return switch state from cache.
+        Run update first."""
+        if self._is_on != 0:
```
Comment (Danielhiversen) [inline switchbot/__init__.py:201]: I think it is better to move this logic to the update function

**Review by Danielhiversen — CHANGES_REQUESTED** [review]
Thanks 👍

File: switchbot/__init__.py [hunk 6]
```diff
@@ -129,6 +162,20 @@ def get_min_time_update(self):
 class Switchbot(SwitchbotDevice):
     """Representation of a Switchbot."""
 
+    def __init__(self, *args, **kwargs) -> None:
+        self._battery_percent = 0
```
Comment (Danielhiversen) [inline switchbot/__init__.py:166]: Move to SwitchbotDevice class

File: switchbot/__init__.py [hunk 7]
```diff
@@ -111,12 +115,41 @@ def _sendcommand(self, key, retry) -> bool:
...
+            return self.get_servicedata(retry - 1, scan_timeout)
+
+        for device in devices:
+            if self._mac.lower() == device.addr.lower():
+                for (adtype, _, value) in device.getScanData():
+                    if adtype == 22:
+                        service_data = value[4:].encode()
+                        service_data = binascii.unhexlify(service_data)
+
+        return service_data
```
Comment (Danielhiversen) [inline switchbot/__init__.py:151]:
```suggestion
                        return service_data

```

File: switchbot/__init__.py (hunk 1 above)
Comment (RenierM26) [reply switchbot/__init__.py:149]: Done. Also added logging.

File: switchbot/__init__.py (hunk 2 above)
Comment (RenierM26) [reply switchbot/__init__.py]: Changed typing to bytearray

File: switchbot/__init__.py (hunk 3 above)
Comment (RenierM26) [reply switchbot/__init__.py]: I agree. Initialize as None

File: switchbot/__init__.py (hunk 4 above)
Comment (RenierM26) [reply switchbot/__init__.py]: Done.

File: switchbot/__init__.py (hunk 5 above)
Comment (RenierM26) [reply switchbot/__init__.py]: Done.

File: switchbot/__init__.py (hunk 6 above)
Comment (RenierM26) [reply switchbot/__init__.py]: Done.

File: switchbot/__init__.py (hunk 7 above)
Comment (RenierM26) [reply switchbot/__init__.py]: Done.

Had to add "return None" to end of method to prevent pylint PEP error. (all should return expression of none should)

**Comment by RenierM26** [comment]
Missed a part of the battery check after moving to main function. Busy fixing.

File: switchbot/__init__.py [hunk 8]
```diff
@@ -129,6 +171,29 @@ def get_min_time_update(self):
...
 
+    def __init__(self, *args, **kwargs) -> None:
+        self._is_on = None
+        self._mode = None
+        super().__init__(*args, **kwargs)
+
+    def update(self, scan_timeout=5) -> None:
+        """Updates the mode, battery percent and state of the device."""
+        barray = self.get_servicedata(scan_timeout=scan_timeout)
+
```
Comment (Danielhiversen) [inline switchbot/__init__.py:186]:
```suggestion
        if barray is None:
            return
```

File: switchbot/__init__.py [hunk 9]
```diff
@@ -111,12 +117,48 @@ def _sendcommand(self, key, retry) -> bool:
...
+            time.sleep(waiting_time)
+        try:
+            devices = bluepy.btle.Scanner().scan(scan_timeout)
+
+        except bluepy.btle.BTLEManagementError:
+            _LOGGER.warning("Error updateing Switchbot.", exc_info=True)
+
+        if retry < 1:
+            _LOGGER.error("Switchbot update failed. Stopping trying.", exc_info=True)
+            return False
```
Comment (Danielhiversen) [inline switchbot/__init__.py:144]:
```suggestion
            return None
```

File: switchbot/__init__.py [hunk 10]
```diff
@@ -175,39 +251,26 @@ def stop(self) -> bool:
...
 
     def update(self, scan_timeout=5) -> None:
-        """Updates the current position, battery percent and light level of the device.
-        Returns after the given timeout period in seconds."""
-        waiting_time = self.get_min_time_update() - time.time()
-        if waiting_time > 0:
-            time.sleep(waiting_time)
-        devices = bluepy.btle.Scanner().scan(scan_timeout)
+        """Updates the current position, battery percent and light level of the device."""
+        barray = self.get_servicedata(scan_timeout=scan_timeout)
```
Comment (Danielhiversen) [inline switchbot/__init__.py:266]:
```suggestion
        if barray is None:
            return
```

File: switchbot/__init__.py [hunk 11]
```diff
@@ -111,12 +117,48 @@ def _sendcommand(self, key, retry) -> bool:
...
+        waiting_time = self._time_between_update_command - time.time()
+        if waiting_time > 0:
+            time.sleep(waiting_time)
+        try:
+            devices = bluepy.btle.Scanner().scan(scan_timeout)
+
+        except bluepy.btle.BTLEManagementError:
+            _LOGGER.warning("Error updateing Switchbot.", exc_info=True)
+
+        if retry < 1:
```
Comment (Danielhiversen) [inline switchbot/__init__.py:142]: This should be moved inside: `    if not devices: `
Now you will return even if you get a response

**Review by Danielhiversen — CHANGES_REQUESTED** [review]

File: switchbot/__init__.py (hunk 9 above)
Comment (RenierM26) [reply switchbot/__init__.py]: Done.

File: switchbot/__init__.py (hunk 10 above)
Comment (RenierM26) [reply switchbot/__init__.py:275]: Done.

File: switchbot/__init__.py (hunk 11 above)
Comment (RenierM26) [reply switchbot/__init__.py]: Done.

File: switchbot/__init__.py (hunk 8 above)
Comment (RenierM26) [reply switchbot/__init__.py:192]: Done.

**Comment by RenierM26** [comment]
Looks like I swapped around the bot "is_on" status. Busy fixing. Just want to test both modes as the toggle mode should also report off (in my opinion).

**Comment by RenierM26** [comment]
Hi @Danielhiversen ,

All local tests pass:

Single Curtain:
1) Light level - Pass
2) Position - Pass
3) Battery level - Pass

Double Curtain Team/Group:
1) Light level - Pass
2) Position - Pass
3) Battery level - Pass

Switch bot in toggle mode:
1) Mode query - pass
2) battery level - pass
3) Always off

Switch bot in switch mode:
1) Mode query - pass
2) battery level - pass
3) On/Off - both tests pass

**Review by Danielhiversen — APPROVED** [review]
Thanks

**Pull request merged**